### PR TITLE
 KEYCLOAK-6541  KEYCLOAK-6745 KEYCLOAK-7267 KEYCLOAK-6738

### PR DESCRIPTION
**Pull request by vramik**

### CODE REVIEW

**Comment by vramik** [comment]
The failure in crossdc group is expected: https://issues.jboss.org/browse/KEYCLOAK-7421

**Comment by tkyjovsk** [comment]
Seems like this introduces some changes to how the testsuite is being executed. The changes should be reflected in the testsuite documentation (README.md and HOW-TO-RUN.md).

**Comment by vramik** [comment]
@tkyjovsk thank you for pointing that out, I've updated mentioned READMEs.

File: testsuite/integration-arquillian/servers/app-server/undertow/src/main/java/org/keycloak/testsuite/arquillian/undertow/container/UndertowAppServerProvider.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import org.keycloak.testsuite.arquillian.container.AppServerContainerSPI;
+import org.keycloak.testsuite.arquillian.undertow.UndertowAppServer;
+
+/**
+ * @author <a href="mailto:vramik@redhat.com">Vlasta Ramik</a>
+ */
+public class UndertowAppServerProvider implements AppServerContainerSPI {
+
+    private Node configuration;
+    private final String containerName = "undertow";
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/servers/app-server/undertow/src/main/java/org/keycloak/testsuite/arquillian/undertow/container/UndertowAppServerProvider.java]: can be static?

File: testsuite/integration-arquillian/servers/app-server/undertow/src/main/java/org/keycloak/testsuite/arquillian/undertow/container/UndertowAppServerProvider.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    private void createChild(String name, String text) {
+        configuration.createChild("property").attribute("name", name).text(text);
+    }
+
+    private Node standaloneContainer() {
+        Node container = new Node("container");
+        container.attribute("mode", "manual");
+        container.attribute("qualifier", "app-server-undertow");
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/servers/app-server/undertow/src/main/java/org/keycloak/testsuite/arquillian/undertow/container/UndertowAppServerProvider.java]: Would be preferred to use constants: `APP_SERVER + "-" + CONTAINER_NAME`

File: testsuite/integration-arquillian/servers/app-server/undertow/src/main/java/org/keycloak/testsuite/arquillian/undertow/container/UndertowAppServerProvider.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        createChild("bindHttpPort", "8280");
+        createChild("adapterImplClass", UndertowAppServer.class.getName());
+
+        return container;
+    }
+
+    private Node haNodeContainer(int number) {
+        Node container = new Node("container");
+        container.attribute("mode", "manual");
+        container.attribute("qualifier", "app-server-undertow-ha-node-" + number);
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/servers/app-server/undertow/src/main/java/org/keycloak/testsuite/arquillian/undertow/container/UndertowAppServerProvider.java]: dtto

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/ContainerConstants.java [hunk 4]
```diff
@@ -14,18 +14,10 @@
...
-    @Ignore
-    public void testAccountManagementSessionsLogout() {
-        //Can't test this because backchannel logout for filter does not invalidate the session
-    }
-    
-}
\ No newline at end of file
+public class ContainerConstants {
+    public static final String APP_SERVER_UNDERTOW = "app-server-undertow";
+    public static final String APP_SERVER_WILDFLY = "app-server-wildfly";
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/ContainerConstants.java]: prefer constants like in previous commit review.

**Review by hmlnarik — COMMENTED** [review]
Great job, @vramik! My comments are mostly nits, so feel free not to implement those.

Just wondering - DeploymentArchiveProcessor could be externalized to the app-server-dependent modules too and would be simplified this way. Is there a reason not to do it here?

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/annotation/AppServerContainers.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+/**
+ * TODO
+ * 
+ */
+@Documented
+@Retention(RUNTIME)
+@Target({ElementType.TYPE})
+public @interface AppServerContainers {
```
[the remark this answers is not in the document]
Comment (vramik) [reply testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/annotation/AppServerContainers.java]: Personally I like it better this way, when enabled containers are specified within standalone annotation.

**Review by hmlnarik — DISMISSED** [review]
Looks good to me but will leave merging only after @mposolda has chance to review it.

**Comment by vramik** [comment]
@mposolda, @hmlnarik the PR is finished and ready for final review. Thank you

**Review by hmlnarik — DISMISSED** [review]
LGTM

**Comment by mposolda** [comment]
- Shouldn't be AppServerContainerSPI renamed to AppServerContainerProvider? There is some small inconsistency as the implementations of the interface are *Provider (For example UndertowAppServerProvider).


- I am not 100% convinced about do all the container configurations programatically (For example I mean all the "createChild" calls in the EAPAppServerProvider.standaloneContainer() ). It has great flexibility, but on the other hand, I am not 100% sure if it's not a bit error prone. In theory we can still have XML files with container definitions in the individual submodules (For example for EAP somewhere inside  testsuite/integration-arquillian/servers/app-server/jboss/eap/src/main/resources/containers/arquillian-eap.xml ). But maybe I am just not used to programatic configuration, not sure... I guess we can keep it for now and see how it works and eventually change later.


- AbstractPhotozExampleAdapterTest.java -- Could we have possibility to declare things like "ContainerAssume.assumeNotAuthServerUndertow();" per class instead of per test methods? I mean have both possibilities available (Something like @Ignore annotation can be declared per class or per single test method)? Or is it already possible? This is also not needed now and shouldn't block this PR, but maybe could be considered for the future.


- My usual question :) - Is it still possible to run adapter tests from Intellij IDEA with Idea JUnit test runner? In other words, when I am on sources for example of the class OIDCPublicKeyRotationAdapterTest.java . I click by right mouse button and then "Run OIDCPublicKeyRotationAdapterTest" or "Debug OIDCPublicKeyRotationAdapterTest" - is this working? Will it use Undertow authServer + Undertow appServer by default?

**Comment by vramik** [comment]
> Shouldn't be AppServerContainerSPI renamed to AppServerContainerProvider? There is some small inconsistency as the implementations of the interface are *Provider (For example UndertowAppServerProvider).

Possibly. If you prefer AppServerContainerProvider I can change that.

> I am not 100% convinced about do all the container configurations programatically (For example I mean all the "createChild" calls in the EAPAppServerProvider.standaloneContainer() ). It has great flexibility, but on the other hand, I am not 100% sure if it's not a bit error prone. In theory we can still have XML files with container definitions in the individual submodules (For example for EAP somewhere inside testsuite/integration-arquillian/servers/app-server/jboss/eap/src/main/resources/containers/arquillian-eap.xml ). But maybe I am just not used to programatic configuration, not sure... I guess we can keep it for now and see how it works and eventually change later.

I was also thinking about that. XML files with definitions are possible, but in that case we need to parse it and it would copy functionality from arquillian. The methods from arq are not public so they cannot be used unless we use Reflection API. It's not good option IMO. But e.g. some king of propetry file or contant class where all propertis would be in single place easily visible. We could discuss it futher, but for now I'd go with current implementation. 

> AbstractPhotozExampleAdapterTest.java -- Could we have possibility to declare things like "ContainerAssume.assumeNotAuthServerUndertow();" per class instead of per test methods? I mean have both possibilities available (Something like @ignore annotation can be declared per class or per single test method)? Or is it already possible? This is also not needed now and shouldn't block this PR, but maybe could be considered for the future.

This is kind of temporary solution. I was thinking about introducing a new annotaion. Based on the annotation we could skip the test automatically in AdapterTestExecutionDecider. But it is possible ignore both single test and whole test class with current implementation. 
`@BeforeClass`
`public static void beforeClass() {`
`	ContainerAssume.assumeNotAuthServerUndertow();`
`}`

> My usual question :) - Is it still possible to run adapter tests from Intellij IDEA with Idea JUnit test runner? In other words, when I am on sources for example of the class OIDCPublicKeyRotationAdapterTest.java . I click by right mouse button and then "Run OIDCPublicKeyRotationAdapterTest" or "Debug OIDCPublicKeyRotationAdapterTest" - is this working? Will it use Undertow authServer + Undertow appServer by default?

I suppose so, by default it's used Undertow authServer + Undertow appServer, but if it works as you described I cannot tell because I don't have Intellij IDEA. If I understood correctly how JUnit test runner works I think it should work.

**Comment by mposolda** [comment]
>   Shouldn't be AppServerContainerSPI renamed to AppServerContainerProvider? There is some small 
> inconsistency as the implementations of the interface are *Provider (For example UndertowAppServerProvider).
> 
> Possibly. If you prefer AppServerContainerProvider I can change that.

Yes, my vote is to change it to AppServerContainerProvider

> I am not 100% convinced about do all the container configurations programatically (For example I mean all the "createChild" calls in the EAPAppServerProvider.standaloneContainer() ). It has great flexibility, but on the other hand, I am not 100% sure if it's not a bit error prone. In theory we can still have XML files with container definitions in the individual submodules (For example for EAP somewhere inside testsuite/integration-arquillian/servers/app-server/jboss/eap/src/main/resources/containers/arquillian-eap.xml ). But maybe I am just not used to programatic configuration, not sure... I guess we can keep it for now and see how it works and eventually change later.
> 
> I was also thinking about that. XML files with definitions are possible, but in that case we need to parse it and it would copy functionality from arquillian. The methods from arq are not public so they cannot be used unless we use Reflection API. It's not good option IMO. But e.g. some king of propetry file or constant class where all properties would be in single place easily visible could be solution. We can discuss it further, but for now I'd go with current implementation.

+1 to discuss it further later and keep the PR as is.

> AbstractPhotozExampleAdapterTest.java -- Could we have possibility to declare things like "ContainerAssume.assumeNotAuthServerUndertow();" per class instead of per test methods? I mean have both possibilities available (Something like @ignore annotation can be declared per class or per single test method)? Or is it already possible? This is also not needed now and shouldn't block this PR, but maybe could be considered for the future.
> 
> This could be kind of temporary solution. I was thinking about introducing a new annotation. Based on the annotation we could skip the test in AdapterTestExecutionDecider. But it is possible ignore both single test and whole test class with current implementation:
> @BeforeClass
> public static void beforeClass() {
> ContainerAssume.assumeNotAuthServerUndertow();
> }

That should be fine, Thanks!

> My usual question :) - Is it still possible to run adapter tests from Intellij IDEA with Idea JUnit test runner? In other words, when I am on sources for example of the class OIDCPublicKeyRotationAdapterTest.java . I click by right mouse button and then "Run OIDCPublicKeyRotationAdapterTest" or "Debug OIDCPublicKeyRotationAdapterTest" - is this working? Will it use Undertow authServer + Undertow appServer by default?
> 
> I suppose so, by default it's used Undertow authServer + Undertow appServer, but if it works as you described I cannot tell because I don't have Intellij IDEA. If I understood correctly how JUnit test runner works I think it should work.

It runs the test with the maven profiles, which are enabled by default. If there is a need to specify maven profile when running the test, then I think it won't easily work. 

In other words, when you run for example this:

```
cd testsuite/integration-arquillian
mvn clean install -Dtest=OIDCPublicKeyRotationAdapterTest
```

will be the test executed by default with the auth-server-undertow + app-server-undertow?

**Comment by vramik** [comment]
> will be the test executed by default with the auth-server-undertow + app-server-undertow?
yes

**Comment by mposolda** [comment]
@vramik Cool, so the only thing is AppServerContainerProvider and we should be ready from my PoV. WDYT?

**Comment by vramik** [comment]
@mposolda as far as I am concerned we are ready. AFAIK @pdrozd wants to take a look as well ;)

**Review by mposolda — DISMISSED** [review]
@vramik Thanks!

**Review by pdrozd — APPROVED** [review]

**Pull request merged**